File: db/firestore.py
import os
import uuid
from google.cloud import firestore
from models.member import Member


_db = None


def get_db() -> firestore.Client:
    global _db
    if _db is None:
        _db = firestore.Client(project=os.getenv("FIRESTORE_PROJECT_ID"))
    return _db


def _tree_ref(user_id: int):
    return get_db().collection("family_trees").document(str(user_id))


def _members_ref(user_id: int):
    return _tree_ref(user_id).collection("members")


# --- Member CRUD ---

def add_member(user_id: int, member: Member) -> Member:
    if not member.id:
        member.id = str(uuid.uuid4())
    _members_ref(user_id).document(member.id).set(member.to_dict())
    return member


def get_member(user_id: int, member_id: str) -> Member | None:
    doc = _members_ref(user_id).document(member_id).get()
    return Member.from_dict(doc.to_dict()) if doc.exists else None


def update_member(user_id: int, member_id: str, fields: dict):
    _members_ref(user_id).document(member_id).update(fields)


def delete_member(user_id: int, member_id: str):
    _members_ref(user_id).document(member_id).delete()


def list_members(user_id: int) -> list[Member]:
    docs = _members_ref(user_id).stream()
    return [Member.from_dict(d.to_dict()) for d in docs]


def search_members(user_id: int, query: str) -> list[Member]:
    query_lower = query.lower()
    return [m for m in list_members(user_id) if query_lower in m.name.lower()]


# --- Approved Users ---

def _approved_ref():
    return get_db().collection("approved_users")


def approve_user(user_id: int, name: str = "", added_by: int = 0):
    from datetime import datetime
    _approved_ref().document(str(user_id)).set({
        "user_id": user_id,
        "name": name,
        "added_by": added_by,
        "approved_at": datetime.utcnow().isoformat(),
    })


def revoke_user(user_id: int):
    _approved_ref().document(str(user_id)).delete()


def is_approved(user_id: int) -> bool:
    from config import ADMIN_IDS
    if user_id in ADMIN_IDS:
        return True
    return _approved_ref().document(str(user_id)).get().exists


def list_approved_users() -> list[dict]:
    return [d.to_dict() for d in _approved_ref().stream()]


# --- Relationship helpers ---

def link_parent_child(user_id: int, parent_id: str, child_id: str):
    parent = get_member(user_id, parent_id)
    child = get_member(user_id, child_id)
    if not parent or not child:
        return

    if child_id not in parent.child_ids:
        parent.child_ids.append(child_id)
        update_member(user_id, parent_id, {"child_ids": parent.child_ids})

    if parent_id not in child.parent_ids:
        child.parent_ids.append(parent_id)
        update_member(user_id, child_id, {"parent_ids": child.parent_ids})


def link_spouses(user_id: int, member_a_id: str, member_b_id: str):
    a = get_member(user_id, member_a_id)
    b = get_member(user_id, member_b_id)
    if not a or not b:
        return

    if member_b_id not in a.spouse_ids:
        a.spouse_ids.append(member_b_id)
        update_member(user_id, member_a_id, {"spouse_ids": a.spouse_ids})

    if member_a_id not in b.spouse_ids:
        b.spouse_ids.append(member_a_id)
        update_member(user_id, member_b_id, {"spouse_ids": b.spouse_ids})


def unlink_parent_child(user_id: int, parent_id: str, child_id: str):
    parent = get_member(user_id, parent_id)
    child = get_member(user_id, child_id)
    if parent and child_id in parent.child_ids:
        parent.child_ids.remove(child_id)
        update_member(user_id, parent_id, {"child_ids": parent.child_ids})
    if child and parent_id in child.parent_ids:
        child.parent_ids.remove(parent_id)
        update_member(user_id, child_id, {"parent_ids": child.parent_ids})


def unlink_spouses(user_id: int, member_a_id: str, member_b_id: str):
    a = get_member(user_id, member_a_id)
    b = get_member(user_id, member_b_id)
    if a and member_b_id in a.spouse_ids:
        a.spouse_ids.remove(member_b_id)
        update_member(user_id, member_a_id, {"spouse_ids": a.spouse_ids})
    if b and member_a_id in b.spouse_ids:
        b.spouse_ids.remove(member_a_id)
        update_member(user_id, member_b_id, {"spouse_ids": b.spouse_ids})

File: frontend/src/lib/firestore.ts
import { randomUUID } from "node:crypto";
import { Firestore } from "@google-cloud/firestore";
import { getFirestoreProjectId } from "./config";
import { memberFromDoc, memberToDoc } from "./member";
import { groupFromDoc, groupToDoc } from "./group";
import type { Group, Member } from "./types";

/**
 * Akses Firestore (port dari `backend/db/firestore.py`).
 *
 * Struktur data:
 *   family_trees/{user_id}/members/{member_id}
 *   family_trees/{user_id}/groups/{group_id}
 *   approved_users/{user_id}
 *
 * `user_id=0` dipakai sebagai pohon bersama keluarga (default).
 */

export const SEARCH_LIMIT = 10;

let _db: Firestore | null = null;

export function getDb(): Firestore {
  if (!_db) {
    _db = new Firestore({ projectId: getFirestoreProjectId() });
  }
  return _db;
}

function _treeRef(userId: number) {
  return getDb().collection("family_trees").doc(String(userId));
}

function _membersRef(userId: number) {
  return _treeRef(userId).collection("members");
}

function _approvedRef() {
  return getDb().collection("approved_users");
}

function _groupsRef(userId: number) {
  return _treeRef(userId).collection("groups");
}

// ── Member CRUD ─────────────────────────────────────────────────────────────

export async function addMember(userId: number, member: Member): Promise<Member> {
  const id = member.id || randomUUID();
  const doc = { ...memberToDoc(member), id };
  await _membersRef(userId).doc(id).set(doc);
  return memberFromDoc(doc);
}

export async function getMember(
  userId: number,
  memberId: string,
): Promise<Member | null> {
  const doc = await _membersRef(userId).doc(memberId).get();
  return doc.exists ? memberFromDoc(doc.data() ?? {}) : null;
}

// ── List / pagination / search ───────────────────────────────────────────────

/** Ambil semua anggota — hanya untuk pohon kecil atau lookup relasi. */
export async function listMembers(userId: number): Promise<Member[]> {
  const snapshot = await _membersRef(userId).orderBy("name_lower").get();
  return snapshot.docs.map((d) => memberFromDoc(d.data()));
}

/** Return `per_page` anggota mulai dari `offset` + info apakah masih ada lagi. */
export async function listMembersPaginated(
  userId: number,
  perPage = 20,
  offset = 0,
): Promise<{ members: Member[]; hasMore: boolean }> {
  const snapshot = await _membersRef(userId)
    .orderBy("name_lower")
    .offset(offset)
    .limit(perPage + 1)
    .get();
  const docs = snapshot.docs;
  const hasMore = docs.length > perPage;
  return {
    members: docs.slice(0, perPage).map((d) => memberFromDoc(d.data())),
    hasMore,
  };
}

/** Jumlah seluruh anggota milik user (untuk pagination & statistik). */
export async function countMembers(userId: number): Promise<number> {
  const snapshot = await _membersRef(userId).count().get();
  return snapshot.data().count;
}

/** Prefix search pada field name_lower; fallback substring untuk data lama. */
export async function searchMembers(
  userId: number,
  query: string,
): Promise<Member[]> {
  const q = query.toLowerCase().trim();
  const snapshot = await _membersRef(userId)
    .orderBy("name_lower")
    .where("name_lower", ">=", q)
    .where("name_lower", "<=", q + "\uf8ff")
    .limit(SEARCH_LIMIT)
    .get();
  let results = snapshot.docs.map((d) => memberFromDoc(d.data()));

  // Fallback: substring search untuk data lama tanpa name_lower.
  if (results.length === 0) {
    const fallback = await _membersRef(userId).limit(200).get();
    results = fallback.docs
      .map((d) => memberFromDoc(d.data()))
      .filter((m) => m.name.toLowerCase().includes(q))
      .slice(0, SEARCH_LIMIT);
  }

  return results;
}

// ── User ↔ Member (Auth0) ────────────────────────────────────────────────────

/** Cari anggota yang akun Auth0-nya (`sub`) tertaut padanya. */
export async function getMemberBySub(
  userId: number,
  sub: string,
): Promise<Member | null> {
  if (!sub) return null;
  const snapshot = await _membersRef(userId)
    .where("auth0_sub", "==", sub)
    .limit(1)
    .get();
  return snapshot.docs.length > 0
    ? memberFromDoc(snapshot.docs[0].data())
    : null;
}

/** Tautkan akun Auth0 (`sub`) ke seorang anggota silsilah. */
export async function linkUserToMember(
  userId: number,
  memberId: string,
  sub: string,
): Promise<void> {
  await updateMember(userId, memberId, { auth0_sub: sub });
}

/**
 * Atur keanggotaan group pada anggota (`group_ids`).
 *
 * Id group yang tidak dikenal dibiarkan ter-filter di layer API; di sini hanya
 * menulis daftar yang sudah tervalidasi.
 */
export async function setMemberGroups(
  userId: number,
  memberId: string,
  groupIds: string[],
): Promise<void> {
  await updateMember(userId, memberId, { group_ids: groupIds });
}

/** Atur status PIC (Person In Charge) pada anggota. */
export async function setMemberPic(
  userId: number,
  memberId: string,
  isPic: boolean,
): Promise<void> {
  await updateMember(userId, memberId, { is_pic: isPic });
}

// ── Group CRUD ───────────────────────────────────────────────────────────────

export async function addGroup(userId: number, group: Group): Promise<Group> {
  const id = group.id || randomUUID();
  const doc = { ...groupToDoc(group), id };
  await _groupsRef(userId).doc(id).set(doc);
  return groupFromDoc(doc);
}

export async function getGroup(
  userId: number,
  groupId: string,
): Promise<Group | null> {
  const doc = await _groupsRef(userId).doc(groupId).get();
  return doc.exists ? groupFromDoc(doc.data() ?? {}) : null;
}

/** Ambil semua grup, diurutkan berdasarkan kode (case-insensitive). */
export async function listGroups(userId: number): Promise<Group[]> {
  const snapshot = await _groupsRef(userId).orderBy("code_lower").get();
  return snapshot.docs.map((d) => groupFromDoc(d.data()));
}

/** Return `perPage` grup mulai dari `offset` + info apakah masih ada lagi. */
export async function listGroupsPaginated(
  userId: number,
  perPage = 20,
  offset = 0,
): Promise<{ groups: Group[]; hasMore: boolean }> {
  const snapshot = await _groupsRef(userId)
    .orderBy("code_lower")
    .offset(offset)
    .limit(perPage + 1)
    .get();
  const docs = snapshot.docs;
  const hasMore = docs.length > perPage;
  return {
    groups: docs.slice(0, perPage).map((d) => groupFromDoc(d.data())),
    hasMore,
  };
}

/** Jumlah seluruh grup milik user (untuk pagination). */
export async function countGroups(userId: number): Promise<number> {
  const snapshot = await _groupsRef(userId).count().get();
  return snapshot.data().count;
}

export async function updateGroup(
  userId: number,
  groupId: string,
  fields: Record<string, unknown>,
): Promise<void> {
  // Jaga code_lower tetap sinkron saat code diubah.
  if (typeof fields.code === "string") {
    fields.code_lower = fields.code.toLowerCase();
  }
  await _groupsRef(userId).doc(groupId).update(fields);
}

export async function deleteGroup(userId: number, groupId: string): Promise<void> {
  await _groupsRef(userId).doc(groupId).delete();
}

// ── Approved Users ───────────────────────────────────────────────────────────

export async function approveUser(
  userId: number,
  name = "",
  addedBy = 0,
): Promise<void> {
  await _approvedRef().doc(String(userId)).set({
    user_id: userId,
    name,
    added_by: addedBy,
    approved_at: new Date().toISOString().replace("Z", ""),
  });
}

export async function revokeUser(userId: number): Promise<void> {
  await _approvedRef().doc(String(userId)).delete();
}

export async function listApprovedUsers(): Promise<Record<string, unknown>[]> {
  const snapshot = await _approvedRef().get();
  return snapshot.docs.map((d) => d.data());
}

export async function updateMember(
  userId: number,
  memberId: string,
  fields: Record<string, unknown>,
): Promise<void> {
  // Jaga name_lower tetap sinkron saat name diubah.
  if (typeof fields.name === "string") {
    fields.name_lower = fields.name.toLowerCase();
  }
  await _membersRef(userId).doc(memberId).update(fields);
}

/** Hapus anggota dan bersihkan semua referensinya dari anggota lain. */
export async function deleteMember(userId: number, memberId: string): Promise<void> {
  await _membersRef(userId).doc(memberId).delete();
  const all = await listMembers(userId);
  for (const m of all) {
    const updates: Record<string, string[]> = {};
    for (const attr of ["parent_ids", "sibling_ids", "spouse_ids", "child_ids"] as const) {
      const ids = m[attr];
      if (ids.includes(memberId)) {
        updates[attr] = ids.filter((id) => id !== memberId);
      }
    }
    if (Object.keys(updates).length > 0) {
      await updateMember(userId, m.id, updates);
    }
  }
}

// ── Relationship helpers ─────────────────────────────────────────────────────

export async function linkParentChild(
  userId: number,
  parentId: string,
  childId: string,
): Promise<void> {
  const [parent, child] = await Promise.all([
    getMember(userId, parentId),
    getMember(userId, childId),
  ]);
  if (!parent || !child) return;
  if (!parent.child_ids.includes(childId)) {
    await updateMember(userId, parentId, {
      child_ids: [...parent.child_ids, childId],
    });
  }
  if (!child.parent_ids.includes(parentId)) {
    await updateMember(userId, childId, {
      parent_ids: [...child.parent_ids, parentId],
    });
  }
}

export async function linkSpouses(
  userId: number,
  memberAId: string,
  memberBId: string,
): Promise<void> {
  const [a, b] = await Promise.all([
    getMember(userId, memberAId),
    getMember(userId, memberBId),
  ]);
  if (!a || !b) return;
  if (!a.spouse_ids.includes(memberBId)) {
    await updateMember(userId, memberAId, {
      spouse_ids: [...a.spouse_ids, memberBId],
    });
  }
  if (!b.spouse_ids.includes(memberAId)) {
    await updateMember(userId, memberBId, {
      spouse_ids: [...b.spouse_ids, memberAId],
    });
  }
}

/**
 * Putuskan relasi orang tua–anak.
 *
 * Relasi dihapus dari kedua arah sehingga hasilnya sama walau urutan
 * `memberAId`/`memberBId` tertukar (mis. A=anak, B=orang tua).
 */
export async function unlinkParentChild(
  userId: number,
  memberAId: string,
  memberBId: string,
): Promise<void> {
  const [a, b] = await Promise.all([
    getMember(userId, memberAId),
    getMember(userId, memberBId),
  ]);
  if (a) {
    const updates: Record<string, string[]> = {};
    if (a.child_ids.includes(memberBId)) {
      updates.child_ids = a.child_ids.filter((id) => id !== memberBId);
    }
    if (a.parent_ids.includes(memberBId)) {
      updates.parent_ids = a.parent_ids.filter((id) => id !== memberBId);
    }
    if (Object.keys(updates).length > 0) {
      await updateMember(userId, memberAId, updates);
    }
  }
  if (b) {
    const updates: Record<string, string[]> = {};
    if (b.child_ids.includes(memberAId)) {
      updates.child_ids = b.child_ids.filter((id) => id !== memberAId);
    }
    if (b.parent_ids.includes(memberAId)) {
      updates.parent_ids = b.parent_ids.filter((id) => id !== memberAId);
    }
    if (Object.keys(updates).length > 0) {
      await updateMember(userId, memberBId, updates);
    }
  }
}

export async function unlinkSpouses(
  userId: number,
  memberAId: string,
  memberBId: string,
): Promise<void> {
  const [a, b] = await Promise.all([
    getMember(userId, memberAId),
    getMember(userId, memberBId),
  ]);
  if (a && a.spouse_ids.includes(memberBId)) {
    await updateMember(userId, memberAId, {
      spouse_ids: a.spouse_ids.filter((id) => id !== memberBId),
    });
  }
  if (b && b.spouse_ids.includes(memberAId)) {
    await updateMember(userId, memberBId, {
      spouse_ids: b.spouse_ids.filter((id) => id !== memberAId),
    });
  }
}

export async function linkSiblings(
  userId: number,
  memberAId: string,
  memberBId: string,
): Promise<void> {
  const [a, b] = await Promise.all([
    getMember(userId, memberAId),
    getMember(userId, memberBId),
  ]);
  if (!a || !b) return;
  if (!a.sibling_ids.includes(memberBId)) {
    await updateMember(userId, memberAId, {
      sibling_ids: [...a.sibling_ids, memberBId],
    });
  }
  if (!b.sibling_ids.includes(memberAId)) {
    await updateMember(userId, memberBId, {
      sibling_ids: [...b.sibling_ids, memberAId],
    });
  }
}

export async function unlinkSiblings(
  userId: number,
  memberAId: string,
  memberBId: string,
): Promise<void> {
  const [a, b] = await Promise.all([
    getMember(userId, memberAId),
    getMember(userId, memberBId),
  ]);
  if (a && a.sibling_ids.includes(memberBId)) {
    await updateMember(userId, memberAId, {
      sibling_ids: a.sibling_ids.filter((id) => id !== memberBId),
    });
  }
  if (b && b.sibling_ids.includes(memberAId)) {
    await updateMember(userId, memberBId, {
      sibling_ids: b.sibling_ids.filter((id) => id !== memberAId),
    });
  }
}

